File: Recursion_and_Backtracking_pepcoding_level_1/2.Recursion_on_array/4.first_index.js
var first_index = (arr, target, i = 0) => {
    if (i === arr.length) return -1;
    let curr = target === arr[i] ? i : -1;
    if (curr != -1) return curr;
    curr = first_index(arr, target, i + 1);
    return curr;
};

console.log(first_index([-5, 34, 342, 34, 32], 5));

////////////////////////////

var first_index = (arr, target, i = 0) => {
    if (i === arr.length) return -1;

    let faith = first_index(arr, target, i + 1);
    return arr[i] == target ? i : faith;
};

console.log(first_index([-5, 34, 342, 34, 32], 342));

// E
// first of( 0 to n)

// F
// first of( 1 to n)

// F + E
// first of (0 , 1 to n)

File: Recursion_and_Backtracking_pepcoding_level_1/2.Recursion_on_array/4.first_index.rb
def first_index arr, target, i = 0
    return -1 if i == arr.size

    ans = first_index(arr, target, i + 1)

    return arr[i] == target ? i : ans
end

# puts first_index [-5, 341, 34, 34, 32], 32

# /////////////////////////////////
def first_index2 arr, target, i = 0
    return -1 if i == arr.size

    return i if arr[i] == target

    first_index2(arr, target, i + 1)
end

# puts first_index2 [-5, 341, 34, 34, 32], 32

# /////////////////////////////////
def first_index3 arr, target, i = 0
    return -1 if i == arr.size

    return arr[i] == target ? i : first_index3(arr, target, i + 1)
end

puts first_index3 [-5, 341, 34, 34, 32], 341
# // E
# // first of( 0 to n)

# // F
# // first of( 1 to n)

# // F + E
# // first of (0 , 1 to n)
